[CostModel] Add a DstTy to getShuffleCost

davemgreen opened the pull request:
A shuffle will take two input vectors and a mask, to produce a new vector of size <MaskElts x SrcEltTy>. Historically it has been assumed that the SrcTy and the DstTy are the same for getShuffleCost, with that being relaxed in recent years. If the Tp passed to getShuffleCost is the SrcTy, then the DstTy can be calculated from the Mask elts and the src elt size, but the Mask is not always provided and the Tp is not reliably always the SrcTy. This has led to situations notably in the SLP vectorizer but also in the generic cost routines where assumption about how vectors will be legalized are built into the generic cost routines - for example whether they will widen or promote, with the cost modelling assuming they will widen but the default lowering to promote for integer vectors.

This patch attempts to start improving that - it originally tried to alter more of the cost model but that too quickly became too many changes at once, so this patch just plumbs in a DstTy to getShuffleCost so that DstTy and SrcTy can be reliably distinguished. The callers of getShuffleCost have been updated to try and include a DstTy that is more accurate. Otherwise it tries to be fairly non-functional, keeping the SrcTy used as the primary type used in shuffle cost routines, only using DstTy where it was in the past (for InsertSubVector for example).

Some asserts have been added that help to check for consistent values when a Mask and a DstTy are provided to getShuffleCost. Some of them took a while to get right, and some non-mask calls might still be incorrect. Hopefully this will provide a useful base to build more shuffles that alter size.

davemgreen commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1360,16 +1360,17 @@ class TargetTransformInfo {
+  getShuffleCost(ShuffleKind Kind, VectorType *DstTy, VectorType *SrcTy,
+                 ArrayRef<int> Mask = {},
                  TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput,
                  int Index = 0, VectorType *SubTp = nullptr,
```
Hello - I believe it will still get used for insertsubreg, if there is not another way to represent the smaller size.

RKSimon commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1360,16 +1360,17 @@ class TargetTransformInfo {
+  /// Mask, or else the array will be empty. The index and subtype parameters
+  /// are used by the subvector insertion and extraction shuffle kinds to show
+  /// the insert/extract point and the type of the subvector being
+  /// inserted/extracted. The operands of the shuffle can be passed through \p
```
Update the description to explain that SubTp should only be used for SK_InsertSubvector?

davemgreen replied:
I've updated the comment (but it might not be true quite yet as we still set and use the SubTp for ExtractSubRegs at the moment).

RKSimon reviewed (approved):
LGTM - please double check the patch summary as I think you've said DstTy instead of SubTy when talking about InsertSubVector

davemgreen committed:
[CostModel] Add a DstTy to getShuffleCost

A shuffle will take two input vectors and a mask, to produce a new vector of
size <MaskElts x SrcEltTy>. Historically it has been assumed that the SrcTy and
the DstTy are the same for getShuffleCost, with that being relaxed in recent
years. If the Tp passed to getShuffleCost is the SrcTy, then the DstTy can be
calculated from the Mask elts and the src elt size, but the Mask is not always
provided and the Tp is not reliably always the SrcTy. This has led to
situations notably in the SLP vectorizer but also in the generic cost routines
where assumption about how vectors will be legalized are built into the generic
cost routines - for example whether they will widen or promote, with the cost
modelling assuming they will widen but the default lowering to promote for
integer vectors.

This patch attempts to start improving that - it originally tried to alter more
of the cost model but that too quickly became too many changes at once, so this
patch just plumbs in a DstTy to getShuffleCost so that DstTy and SrcTy can be
reliably distinguished. The callers of getShuffleCost have been updated to try
and include a DstTy that is more accurate. Otherwise it tries to be fairly
non-functional, keeping the SrcTy used as the primary type used in shuffle
cost routines, only using DstTy where it was in the past (for InsertSubVector
for example).

Some asserts have been added that help to check for consistent values when a
Mask and a DstTy are provided to getShuffleCost. Some of them took a while to
get right, and some non-mask calls might still be incorrect. Hopefully this
will provide a useful base to build more shuffles that alter size.

davemgreen commented:
Thanks. I have rebased this and given it some more testing on various test suites and architectures. It is intended to roughly be an NFC, but there is a chance some of the new asserts will trigger in some cases, with mismatching dst types and masks.
